File: Java/Composite/Gemini 1.5 Flash/round1/code.java
import java.util.HashMap;
import java.util.Map;

// Abstract Component (ATM component)
abstract class ATMComponent {
    protected String name;

    public ATMComponent(String name) {
        this.name = name;
    }

    public abstract void execute();

    public abstract String getName();
}

// Leaf Component (Basic ATM operations)
class BalanceDisplay extends ATMComponent {
    private Account account;

    public BalanceDisplay(Account account) {
        super("Balance Display");
        this.account = account;
    }

    @Override
    public void execute() {
        System.out.println("Your current balance is: " + account.getBalance());
    }

    @Override
    public String getName() {
        return name;
    }
}

class CashWithdrawal extends ATMComponent {
    private Account account;

    public CashWithdrawal(Account account) {
        super("Cash Withdrawal");
        this.account = account;
    }

    @Override
    public void execute() {
        double amount = getWithdrawalAmount();
        if (account.getBalance() >= amount) {
            account.withdraw(amount);
            System.out.println("Cash withdrawn successfully. Your new balance is: " + account.getBalance());
        } else {
            System.out.println("Insufficient funds!");
        }
    }

    @Override
    public String getName() {
        return name;
    }

    // Mock method for getting withdrawal amount (replace with user input)
    private double getWithdrawalAmount() {
        return 100.00;
    }
}

class Deposit extends ATMComponent {
    private Account account;

    public Deposit(Account account) {
        super("Deposit");
        this.account = account;
    }

    @Override
    public void execute() {
        double amount = getDepositAmount();
        account.deposit(amount);
        System.out.println("Deposit successful. Your new balance is: " + account.getBalance());
    }

    @Override
    public String getName() {
        return name;
    }

    // Mock method for getting deposit amount (replace with user input)
    private double getDepositAmount() {
        return 200.00;
    }
}

class PINChange extends ATMComponent {
    private Account account;

    public PINChange(Account account) {
        super("PIN Change");
        this.account = account;
    }

    @Override
    public void execute() {
        String newPIN = getNewPIN();
        if (newPIN != null) {
            account.setPIN(newPIN);
            System.out.println("PIN changed successfully.");
        }
    }

    @Override
    public String getName() {
        return name;
    }

    // Mock method for getting new PIN (replace with user input)
    private String getNewPIN() {
        return "1234";
    }
}

// Composite Component (ATM Menu)
class ATMMenu extends ATMComponent {
    private Map<String, ATMComponent> operations;

    public ATMMenu() {
        super("ATM Menu");
        operations = new HashMap<>();
    }

    public void addOperation(ATMComponent operation) {
        operations.put(operation.getName(), operation);
    }

    @Override
    public void execute() {
        System.out.println("ATM Menu:");
        for (String operationName : operations.keySet()) {
            System.out.println(operationName);
        }

        String choice = getMenuChoice();
        if (operations.containsKey(choice)) {
            operations.get(choice).execute();
        } else {
            System.out.println("Invalid choice!");
        }
    }

    @Override
    public String getName() {
        return name;
    }

    // Mock method for getting user menu choice (replace with user input)
    private String getMenuChoice() {
        return "Cash Withdrawal";
    }
}

// Account class (Simulates account data)
class Account {
    private double balance;
    private String PIN;

    public Account(double balance, String PIN) {
        this.balance = balance;
        this.PIN = PIN;
    }

    public double getBalance() {
        return balance;
    }

    public void withdraw(double amount) {
        balance -= amount;
    }

    public void deposit(double amount) {
        balance += amount;
    }

    public String getPIN() {
        return PIN;
    }

    public void setPIN(String newPIN) {
        this.PIN = newPIN;
    }
}

// Example usage:
public class ATMSystem {
    public static void main(String[] args) {
        Account account = new Account(1000.00, "1111");
        ATMMenu menu = new ATMMenu();
        menu.addOperation(new BalanceDisplay(account));
        menu.addOperation(new CashWithdrawal(account));
        menu.addOperation(new Deposit(account));
        menu.addOperation(new PINChange(account));
        menu.execute();
    }
}

File: Python/Composite/Gemini 1.5 Flash/round1/code.py
from abc import ABC, abstractmethod

class ATMComponent(ABC):
    @abstractmethod
    def process(self, user):
        pass

class Authentication(ATMComponent):
    def __init__(self, max_attempts=3):
        self.max_attempts = max_attempts
        self.attempts = 0

    def process(self, user):
        if self.attempts >= self.max_attempts:
            raise ValueError("Card seized due to exceeding attempts.")
        
        # Simulate card and PIN verification
        if user.card_number == "1234567890" and user.pin == "1234":
            print("Authentication successful!")
            self.attempts = 0
            return True
        else:
            self.attempts += 1
            print("Authentication failed. Please try again.")
            return False

class BalanceInquiry(ATMComponent):
    def process(self, user):
        print(f"Your current balance is: {user.balance}")

class Withdrawal(ATMComponent):
    def process(self, user):
        amount = int(input("Enter withdrawal amount: "))
        if amount > user.balance:
            print("Insufficient funds!")
        else:
            user.balance -= amount
            print(f"Withdrawal successful. New balance: {user.balance}")

class Deposit(ATMComponent):
    def process(self, user):
        amount = int(input("Enter deposit amount: "))
        user.balance += amount
        print(f"Deposit successful. New balance: {user.balance}")

class PINChange(ATMComponent):
    def process(self, user):
        old_pin = input("Enter old PIN: ")
        if old_pin != user.pin:
            print("Incorrect old PIN.")
            return
        
        new_pin1 = input("Enter new PIN: ")
        new_pin2 = input("Confirm new PIN: ")
        if new_pin1 == new_pin2:
            user.pin = new_pin1
            print("PIN changed successfully.")
        else:
            print("PINs do not match.")

class PrintReceipt(ATMComponent):
    def process(self, user):
        print("Printing receipt...")

class ATMComposite(ATMComponent):
    def __init__(self):
        self.components = []

    def add(self, component):
        self.components.append(component)

    def remove(self, component):
        self.components.remove(component)

    def process(self, user):
        for component in self.components:
            component.process(user)

class User:
    def __init__(self, card_number, pin, balance):
        self.card_number = card_number
        self.pin = pin
        self.balance = balance

def main():
    user = User("1234567890", "1234", 1000)

    atm = ATMComposite()
    atm.add(Authentication())
    atm.add(BalanceInquiry())
    atm.add(Withdrawal())
    atm.add(Deposit())
    atm.add(PINChange())
    atm.add(PrintReceipt())

    while True:
        print("\nATM Menu:")
        print("1. Authentication")
        print("2. Balance Inquiry")
        print("3. Withdrawal")
        print("4. Deposit")
        print("5. Change PIN")
        print("6. Print Receipt")
        print("7. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            if atm.components[0].process(user):
                continue
        elif choice == '2':
            atm.components[1].process(user)
        elif choice == '3':
            atm.components[2].process(user)
        elif choice == '4':
            atm.components[3].process(user)
        elif choice == '5':
            atm.components[4].process(user)
        elif choice == '6':
            atm.components[5].process(user)
        elif choice == '7':
            break
        else:
            print("Invalid choice.")

if __name__ == "__main__":
    main()